### PR TITLE
[CARBONDATA-3968]Added test cases for hive read complex types and handled other issues

### PR DESCRIPTION
**Pull request by akkio-97**
### Why is this PR needed?
1) Handled issues related to reading of byte, varchar and decimal types.
2) Map of primitive type with only one row inserted had issues.
3) Added test cases for hive read complex types.
 
 ### What changes were proposed in this PR?
**Handled the following-**
1. Byte datatype check was not present.
2. Varchar and decimal types were causing a problem during conversion from hive datatypes to carbon(due to the presence of parentheses).
3. Handled map parser issue.
4. Added test cases for all primitive types excluding -  binary datatype.

 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2157/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3898/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3909/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2168/

File: integration/hive/src/test/java/org/apache/carbondata/hive/HiveTestUtils.java [hunk 1]
```diff
@@ -65,7 +74,12 @@ public boolean checkAnswer(ResultSet actual, ResultSet expected) throws SQLExcep
       Assert.assertTrue(numOfColumnsExpected > 0);
       Assert.assertEquals(actual.getMetaData().getColumnCount(), numOfColumnsExpected);
       for (int i = 1; i <= numOfColumnsExpected; i++) {
-        Assert.assertEquals(actual.getString(i), actual.getString(i));
+        if (actual.getString(i).contains(":")) {
+          Assert.assertTrue(checkMapPairsIgnoringOrder(actual.getString(i), expected.getString(i)));
+        } else {
+          Assert.assertEquals(actual.getString(i), expected.getString(i));
+        }
+        // System.out.println(actual.getString(i));
```
Comment (vikramahuja1001) [inline integration/hive/src/test/java/org/apache/carbondata/hive/HiveTestUtils.java:82]: Remove this comment

**Comment by vikramahuja1001** [comment]
Add jira ID

File: integration/hive/src/main/java/org/apache/carbondata/hive/util/DataTypeUtil.java [hunk 2]
```diff
@@ -64,13 +69,23 @@ public static DataType convertHiveTypeToCarbon(String type) throws SQLException
       return DataTypes.createArrayType(convertHiveTypeToCarbon(subType));
     } else if (type.startsWith("map<")) {
       String[] subType = (type.substring(type.indexOf("<") + 1, type.indexOf(">"))).split(",");
+      for (int i = 0; i < subType.length; i++) {
+        if (subType[i].startsWith("decimal")) {
+          subType[i] += ',' + subType[++i];
```
Comment (vikramahuja1001) [inline integration/hive/src/main/java/org/apache/carbondata/hive/util/DataTypeUtil.java:74]: Use CarbonCommonConstants.COMMA instead of ','

File: integration/hive/src/main/java/org/apache/carbondata/hive/util/DataTypeUtil.java [hunk 3]
```diff
@@ -64,13 +69,23 @@ public static DataType convertHiveTypeToCarbon(String type) throws SQLException
...
           .createMapType(convertHiveTypeToCarbon(subType[0]), convertHiveTypeToCarbon(subType[1]));
     } else if (type.startsWith("struct<")) {
       String[] subTypes =
           (type.substring(type.indexOf("<") + 1, type.indexOf(">"))).split(",");
       List<StructField> structFieldList = new ArrayList<>();
-      for (String subType : subTypes) {
+      for (int i = 0; i < subTypes.length; i++) {
+        String subType = subTypes[i];
+        if (subType.startsWith("decimal")) {
+          subType += ',' + subTypes[++i];
```
Comment (vikramahuja1001) [inline integration/hive/src/main/java/org/apache/carbondata/hive/util/DataTypeUtil.java:87]: Use CarbonCommonConstants.COMMA

File: integration/hive/src/test/java/org/apache/carbondata/hive/HiveTestUtils.java (hunk 1 above)
Comment (akkio-97) [reply integration/hive/src/test/java/org/apache/carbondata/hive/HiveTestUtils.java]: done

File: integration/hive/src/main/java/org/apache/carbondata/hive/util/DataTypeUtil.java (hunk 2 above)
Comment (akkio-97) [reply integration/hive/src/main/java/org/apache/carbondata/hive/util/DataTypeUtil.java]: done

File: integration/hive/src/main/java/org/apache/carbondata/hive/util/DataTypeUtil.java (hunk 3 above)
Comment (akkio-97) [reply integration/hive/src/main/java/org/apache/carbondata/hive/util/DataTypeUtil.java]: done

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2198/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3938/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2228/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3987/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2247/

**Comment by akkio-97** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2248/

File: processing/src/main/java/org/apache/carbondata/processing/loading/parser/impl/MapParserImpl.java [hunk 4]
```diff
@@ -73,9 +73,12 @@ public ArrayObject parse(Object data) {
 
   @Override
   public ArrayObject parseRaw(Object data) {
-    Object keyArray = ((Object[]) data)[0];
-    Object valueArray = ((Object[]) data)[1];
-    return new ArrayObject(new Object[]{child.parseRaw(keyArray), child.parseRaw(valueArray)});
+    Object[] keyValuePairs = ((Object[]) data);
```
Comment (kunal642) [inline processing/src/main/java/org/apache/carbondata/processing/loading/parser/impl/MapParserImpl.java:76]: What is the reason for this change?

File: integration/hive/src/main/java/org/apache/carbondata/hive/util/DataTypeUtil.java [hunk 5]
```diff
@@ -21,25 +21,31 @@
...
 
+import org.apache.commons.lang.ArrayUtils;
+
 public class DataTypeUtil {
 
   public static DataType convertHiveTypeToCarbon(String type) throws SQLException {
     if ("string".equalsIgnoreCase(type) || type.startsWith("char")) {
       return DataTypes.STRING;
-    } else if ("varchar".equalsIgnoreCase(type)) {
+    } else if (!type.startsWith("map<") && !type.startsWith("array<") && !type.startsWith("struct<")
```
Comment (kunal642) [inline integration/hive/src/main/java/org/apache/carbondata/hive/util/DataTypeUtil.java:36]: why map, struct, array check required?

File: processing/src/main/java/org/apache/carbondata/processing/loading/parser/impl/MapParserImpl.java (hunk 4 above)
Comment (akkio-97) [reply processing/src/main/java/org/apache/carbondata/processing/loading/parser/impl/MapParserImpl.java:76]: Here data is a list of key-value pairs. So both the variables keyArray and valueArray contained (key,value) pairs in each which was wrong. So made the required changes.

File: integration/hive/src/main/java/org/apache/carbondata/hive/util/DataTypeUtil.java (hunk 5 above)
Comment (akkio-97) [reply integration/hive/src/main/java/org/apache/carbondata/hive/util/DataTypeUtil.java]: made required changes, they are not required

File: processing/src/main/java/org/apache/carbondata/processing/loading/parser/impl/MapParserImpl.java (hunk 4 above)
Comment (kunal642) [reply processing/src/main/java/org/apache/carbondata/processing/loading/parser/impl/MapParserImpl.java:76]: okay

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3998/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2258/

**Comment by kunal642** [comment]
LGTM
